File: Algorithm/kahn.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <stack>

class Graph {
public:
    int vertices;
    std::vector<std::vector<int>> adjacencyList;

    Graph(int v) : vertices(v), adjacencyList(v) {}

    void addEdge(int u, int v) {
        adjacencyList[u].push_back(v);
    }

    void topologicalSortKahn() {
        std::vector<int> inDegree(vertices, 0);

        for (int i = 0; i < vertices; ++i) {
            for (int neighbor : adjacencyList[i]) {
                inDegree[neighbor]++;
            }
        }

        std::queue<int> q;
        for (int i = 0; i < vertices; ++i) {
            if (inDegree[i] == 0) {
                q.push(i);
            }
        }

        std::vector<int> result;
        while (!q.empty()) {
            int u = q.front();
            q.pop();
            result.push_back(u);

            for (int neighbor : adjacencyList[u]) {
                inDegree[neighbor]--;
                if (inDegree[neighbor] == 0) {
                    q.push(neighbor);
                }
            }
        }

        std::cout << "Topological order (Kahn's Algorithm): ";
        for (int node : result) {
            std::cout << node << " ";
        }
        std::cout << std::endl;
    }

    void topologicalSortDFS() {
        std::vector<bool> visited(vertices, false);
        std::stack<int> st;

        for (int i = 0; i < vertices; ++i) {
            if (!visited[i]) {
                topologicalSortDFSUtil(i, visited, st);
            }
        }

        std::cout << "Topological order (DFS): ";
        while (!st.empty()) {
            std::cout << st.top() << " ";
            st.pop();
        }
        std::cout << std::endl;
    }

private:
    void topologicalSortDFSUtil(int u, std::vector<bool>& visited, std::stack<int>& st) {
        visited[u] = true;

        for (int neighbor : adjacencyList[u]) {
            if (!visited[neighbor]) {
                topologicalSortDFSUtil(neighbor, visited, st);
            }
        }

        st.push(u);
    }
};

int main() {
    Graph g(6);
    g.addEdge(5, 2);
    g.addEdge(5, 0);
    g.addEdge(4, 0);
    g.addEdge(4, 1);
    g.addEdge(2, 3);
    g.addEdge(3, 1);

    std::cout << "Graph edges: 5->2, 5->0, 4->0, 4->1, 2->3, 3->1" << std::endl;
    g.topologicalSortKahn();
    g.topologicalSortDFS();

    return 0;
}

File: Algorithm/kahn.py
from collections import deque, defaultdict


class Graph:
    def __init__(self, vertices):
        self.vertices = vertices
        self.adjacency_list = defaultdict(list)

    def add_edge(self, u, v):
        self.adjacency_list[u].append(v)

    def topological_sort_kahn(self):
        in_degree = [0] * self.vertices

        for i in range(self.vertices):
            for neighbor in self.adjacency_list[i]:
                in_degree[neighbor] += 1

        queue = deque()
        for i in range(self.vertices):
            if in_degree[i] == 0:
                queue.append(i)

        result = []
        while queue:
            u = queue.popleft()
            result.append(u)

            for neighbor in self.adjacency_list[u]:
                in_degree[neighbor] -= 1
                if in_degree[neighbor] == 0:
                    queue.append(neighbor)

        print("Topological order (Kahn's Algorithm):", ' '.join(map(str, result)))

    def topological_sort_dfs(self):
        visited = [False] * self.vertices
        stack = []

        def dfs_util(u):
            visited[u] = True

            for neighbor in self.adjacency_list[u]:
                if not visited[neighbor]:
                    dfs_util(neighbor)

            stack.append(u)

        for i in range(self.vertices):
            if not visited[i]:
                dfs_util(i)

        result = stack[::-1]
        print("Topological order (DFS):", ' '.join(map(str, result)))


g = Graph(6)
g.add_edge(5, 2)
g.add_edge(5, 0)
g.add_edge(4, 0)
g.add_edge(4, 1)
g.add_edge(2, 3)
g.add_edge(3, 1)

print("Graph edges: 5->2, 5->0, 4->0, 4->1, 2->3, 3->1")
g.topological_sort_kahn()
g.topological_sort_dfs()
